### ISSUE TITLE
Meta globs in config.js not respecting file extension when explicit folder used in module id

### ISSUE THREAD

**Issue by duke9509** [systemjs/systemjs#915]
In order to circumvent https://github.com/systemjs/systemjs/issues/450, I attempted to use the `scriptLoad` meta property, like so:
```
meta: {
  "views/*.js": {scriptLoad: true}
}
```
where `views/` is a folder that contains both JavaScript and HTML components of an Aurelia project.

This led to the html files in `views/` appearing in the `src` attributes of injected `<script>` tags.

Reproducible with the following steps:

1. Create a new project with jspm/SystemJS
2. Create a folder called `foo/` in your base directory
3. Create a `bar.html` file in `foo/` and give it at least a basic structure (`DOCTYPE`, `<html></html>`)
4. In your `config.js`, add the following:

    ```
    meta: {
      "foo/*.js": {scriptLoad: true}
    }
    ```

5. Create an `index.html` in your base directory with the following content:

    ```
    <!DOCTYPE HTML>
    <html>
      <head>
        <script src="jspm_packages/system.js"></script>
        <script src="config.js"></script>
        <script>
          System.import('foo/bar.html!text');
        </script>
      </head>
      <body></body>
    </html>
    ```

6. Load the page and note the following in the `<head>` of the DOM (with accompanying syntax error in the console):
`<script src="http://localhost:9000/foo/bar.html" async=""></script>`

However, this seems to only happen when the string passed to `System.import()` has a folder in it. See the following steps:

7. In the same project, now go back to `config.js` and add the following:

    ```
    paths: {
      "*": "foo/*"
    }
    ```

8. Go back to `index.html` and remove the `foo/` from the `System.import()` call.
9. Load the page again and see that the error is gone.

**Comment by guybedford** [comment]
Yes, source based transformations like plugins are not compatible with the `scriptLoad: true` metadata option. You should exclude these from the `scriptLoad: true` setting.

**Comment by duke9509** [comment]
I get that. That's not the issue. I was trying to exclude the affected files from the setting.
The issue is that excluding them didn't work.
The glob `"views/*.js"` was applying the setting to files that did not, in fact, end in `.js`. Is this somehow intended behavior?

**Comment by duke9509** [comment]
Realized I didn't exactly title this thing clearly. Tried to focus it a bit more.

**Comment by guybedford** [comment]
Thanks, this sounds like a bug then. Does it help if you configure the plugin via meta itself:

```javascript
meta: {
  "views/*.html": {loader: 'text'}
}
```

with `System.import('views/bar.html')`?

**Comment by duke9509** [comment]
Even with that meta property set, `System.import("foo/bar.html")` throws a 404 looking for `foo/bar.html.js`. It seems to stick a .js extension on the file automatically if there's no plugin.

**Comment by guybedford** [comment]
Ah is this with `defaultJSExtensions: true`? Sounds like that is the underlying cause here. This property will be deprecated though in the next major, so I assume the bug will go with it.

The way to work around it would be this config then:

```javascript
packages: {
  'foo': {
    defaultExtension: false,
    modules: {
      '*.html': { loader: 'text' },
      '*.js': { scriptLoad: true }
    }
  }
}
```

**Comment by duke9509** [comment]
Ah, yes that setting is set! I didn't even notice it there. Either something set it automatically or I completely forgot putting it there.

I'll try that config.

**Comment by duke9509** [comment]
Still throws the same 404 with that `packages` config.

**Comment by OliverJAsh** [comment]
I've also ran into this bug also, trying to use the following:

``` js
  meta: {
    "*.json": {
      loader: "json"
    }
  },
```

@duke9509 Were you able to get this working?

**Comment by OliverJAsh** [comment]
@guybedford Using the `packages` config, is it possible to use default JS extensions for all files except those ending in `.json`? E.g.:

``` js
  packages: {
    'shared': {
      defaultExtension: 'js',
      modules: {
        '*.json': {
            // Don't use default extension?
            loader: 'json'
        }
      }
    }
  },
```

**Comment by duke9509** [comment]
@OliverJAsh No, I didn't. I just worked around it by rearranging my project so that all of my JavaScript files were in folders that only contained JavaScript.

**Comment by guybedford** [comment]
@OliverJAsh the above does work - modules values inform opt-out of default extension.

**Comment by guybedford** [comment]
I believe this was fixed in https://github.com/systemjs/systemjs/commit/41d08addaf6b8f5760d5e1d55b346c86c99bf98b.

**Comment by guybedford** [comment]
Released in 0.19.7.

**Comment by maybeec** [comment]
@OliverJAsh, @guybedford I have seen this quite old issue here and wanted to get exactly the behavior @OliverJAsh requested. However, with systemJS 0.19.40 am an not able to get it run. I have such kind of configuration for my own package:

```js
packages: {
      'app': {
        main: './main.js',
        defaultExtension: 'js',
        modules: {
          '*!text': {
            defaultExtension: false,
            loader: 'text'
          }
        }
      }
    }
```

However, retrieving a json file by `import subNavConfig from './navigation.config.json!text'` I will get ` subNavConfig` to be `undefined`. The systemjs-text plugin is working as it already handles the angular templates. What am I doing wrong? I do not even get a 404. So the file seems to be resolved but not read successfully. I can even see the json file is served and I can even read it with `require(PATH)`.
